CppCustomVisualizer2/dll/RootVisualizer.cpp

```diff
@@ -0,0 +1,379 @@
+    _Out_ DkmArray<DkmChildVisualizedExpression*>* pItems
+)
+{
+    HRESULT hr = S_OK;
+
+    CComPtr<DkmPointerValueHome> pPointerValueHome = DkmPointerValueHome::TryCast(pVisualizedExpression->ValueHome());
+
+    CAtlList<CComPtr<DkmChildVisualizedExpression>> childItems;
+
+    for (UINT32 i = StartIndex; i < Count + StartIndex && i < m_size; i++)
+    {
+        CComPtr<DkmPointerValueHome> pParentPointerValueHome = DkmPointerValueHome::TryCast(pVisualizedExpression->ValueHome());
+
+        CComPtr<DkmString> pChildName;
+        hr = DkmString::Create(DkmSourceString(L"[Index]"), &pChildName);
+        if (FAILED(hr))
+        {
+            return hr;
+        }
+
+        CComPtr<DkmString> pChildFullName;
+        hr = m_pVisualizedExpression->CreateDefaultChildFullName(0, &pChildFullName);
+        if (FAILED(hr))
+        {
+            return hr;
+        }
+
+        CComObject<CChildVisualizer>* pChildVisualizer;
+        CComObject<CChildVisualizer>::CreateInstance(&pChildVisualizer);
+        pChildVisualizer->Initialize(m_pVisualizedExpression, m_size, i, m_fIsPointer);
+
+        CComPtr<DkmEvaluationResult> pEvaluationResult;
+        hr = pChildVisualizer->CreateEvaluationResult(
+            pChildName,
+            pChildFullName,
+            nullptr,
+            DkmRootVisualizedExpressionFlags::None,
+            m_pVisualizedExpression,
+            m_pVisualizedExpression->InspectionContext(),
+            i,
+            &pEvaluationResult
+        );
+        if (FAILED(hr))
+        {
+            return hr;
+        }
+
+        CComPtr<DkmChildVisualizedExpression> pChildVisualizedExpression;
+        hr = DkmChildVisualizedExpression::Create(
+            m_pVisualizedExpression->InspectionContext(),
+            m_pVisualizedExpression->VisualizerId(),
+            m_pVisualizedExpression->SourceId(),
+            m_pVisualizedExpression->StackFrame(),
+            pPointerValueHome,
+            pEvaluationResult,
+            m_pVisualizedExpression,
+            i,
+            pChildVisualizer,
+            &pChildVisualizedExpression
+        );
+        if (FAILED(hr))
+        {
+            return hr;
+        }
+
+        childItems.AddTail(pChildVisualizedExpression);
+    }
+
+    CAutoDkmArray<DkmChildVisualizedExpression*> resultValues;
+    DkmAllocArray(childItems.GetCount(), &resultValues);
+    if (FAILED(hr))
+    {
+        return hr;
+    }
+
+    UINT32 j = 0;
+    POSITION pos = childItems.GetHeadPosition();
+    while (pos != NULL)
+    {
+        CComPtr<DkmChildVisualizedExpression> pCurr = childItems.GetNext(pos);
+        resultValues.Members[j] = pCurr.Detach();
+        j++;
+    }
+
+    *pItems = resultValues.Detach();
+
+    return hr;
+}
+
+HRESULT CRootVisualizer::GetSize(
+    _In_ Evaluation::DkmVisualizedExpression* pVisualizedExpression,
+    _In_ DkmString* pFullName,
+    _In_ LPCWSTR pMemberName,
+    _In_ bool rootIsPointer,
+    _Out_ size_t* pSize
```

Comment:
>size_t [](http://example.com/codeflow?start=10&length=6)

I think I would suggest avoiding `size_t` here since the `size_t` in the target process can be different from the EE. So I would pick something that will not very.

Reply:
How about `unsigned long long`? So we make sure it's large enough?

Reply:
That works.